message/validation/roundthresholds/round_thresholds.go

```diff
@@ -0,0 +1,87 @@
+package roundthresholds
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	specqbft "github.com/bloxapp/ssv-spec/qbft"
+	spectypes "github.com/bloxapp/ssv-spec/types"
+
+	beaconprotocol "github.com/bloxapp/ssv/protocol/v2/blockchain/beacon"
+	"github.com/bloxapp/ssv/protocol/v2/qbft/roundtimer"
+)
+
+// Cache contains thresholds when each round finishes for each role.
+type Cache struct {
+	bn         beaconprotocol.BeaconNetwork
+	thresholds map[spectypes.BeaconRole][]time.Duration
+	mu         sync.Mutex
+}
+
+func New(bn beaconprotocol.BeaconNetwork) *Cache {
+	return &Cache{
+		bn:         bn,
+		thresholds: make(map[spectypes.BeaconRole][]time.Duration),
+		mu:         sync.Mutex{},
+	}
+}
+
+// InitThresholds fills threshold cache for given role.
+func (c *Cache) InitThresholds(role spectypes.BeaconRole) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	unusedCtx := context.Background()
+	rt := roundtimer.New(unusedCtx, c.bn, role, nil)
+
+	var cumulativeDuration time.Duration
+	round := specqbft.Round(1)
+	c.thresholds[role] = []time.Duration{}
+
+	for {
+		roundDuration := rt.RoundDuration(round)
+		if roundDuration <= 0 {
+			panic(fmt.Sprintf("invalid round duration for round %d", round))
```

Comment:
don't panic, either return error or use logger to fatal if this is a really fatal case. can this never happen ?

Reply:
Round duration must never be negative/zero, so this case would mean that there's a critical issue in the round timer that must be noticed immediately, so panic fits best in my opinion. I agree that fatal as a workaround for panic would be better